### PR TITLE
723 Git Metrics in Diff City

### PR DESCRIPTION
**Pull request by Omidious**
We want to be able to see Git metrics like commit frequency, number of code added and deleted, etc. in Diff City when we generate a city. Use GitGraphProvider in Diff City in combination with VCSGraphProvider to generate Git metrics. Enter the required information under the VCS tab.

### CODE REVIEW

**Comment by falko17** [comment]
Before I review this, how does this relate to #730? Does it directly depend on it, i.e., should #730 be merged first?

**Comment by Omidious** [comment]
> Before I review this, how does this relate to #730? Does it directly depend on it, i.e., should #730 be merged first?
GitGraphProvider uses VCSGraphProvider in #730 to add Git metrics to a graph generated from a Git repository via VCSGraphProvider. I think it would be better if #730 was merged first.

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+
+namespace SEE.GraphProviders
+{
+    /// <summary>
+    /// Calculates metrics between 2 revisions from a git repository and adds these to a graph.
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:15]: Numbers until (including) twelve are generally given in words. So, "2" should be replaced by "two".

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+        /// </summary>
+        private string OldRevision = string.Empty;
+
+        /// <summary>
+        /// The newer revision against which the <see cref="OldRevision"/> is to be compared.
+        /// </summary>
+        private string NewRevision = string.Empty;
+
+        /// <summary>
+        /// Calculates metrics between 2 revisions from a git repository and adds these to <paramref name="graph"/>.
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:36]: Numbers until (including) twelve are generally given in words. So, "2" should be replaced by "two".

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+        /// Calculates the number of lines of code added and deleted for each file changed between two commits and adds them as metrics to <paramref name="graph"/>.
+        /// <param name="graph">an existing graph where to add the metrics</param>
+        /// <param name="vcsPath">the path to the VCS containing the two revisions to be compared</param>
+        /// <param name="oldRevision">the older revision that constitutes the baseline of the comparison</param>
+        /// <param name="newRevision">the newer revision against which the <see cref="oldRevision"/> is to be compared</param>
+        protected void AddLineofCodeChurnMetric(Graph graph, string vcsPath, string oldRevision, string newRevision)
+        {
+            using (var repo = new Repository(vcsPath))
+            {
+                var oldCommit = repo.Lookup<Commit>(oldRevision);
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:158]: [Do not use `var`.](https://github.com/uni-bremen-agst/SEE/wiki/ReviewChecklist)

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+                var newCommit = repo.Lookup<Commit>(newRevision);
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:159]: [Do not use `var`.](https://github.com/uni-bremen-agst/SEE/wiki/ReviewChecklist)

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+
+                var changes = repo.Diff.Compare<Patch>(oldCommit.Tree, newCommit.Tree);
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:161]: [Do not use `var`.](https://github.com/uni-bremen-agst/SEE/wiki/ReviewChecklist)

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+
+                foreach (var change in changes)
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:163]: [Do not use `var`.](https://github.com/uni-bremen-agst/SEE/wiki/ReviewChecklist)

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+                {
+                    foreach (var node in graph.Nodes())
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:165]: [Do not use `var`.](https://github.com/uni-bremen-agst/SEE/wiki/ReviewChecklist)

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+        /// <summary>
+        /// Calculates the number of unique developers who contributed to each file for each file changed between two commits and adds it as a metric to <paramref name="graph"/>.
+        /// </summary>
+        /// <param name="graph">an existing graph where to add the metrics</param>
+        /// <param name="vcsPath">the path to the VCS containing the two revisions to be compared</param>
+        /// <param name="oldRevision">the older revision that constitutes the baseline of the comparison</param>
+        /// <param name="newRevision">the newer revision against which the <see cref="oldRevision"/> is to be compared</param>
+        protected void AddNumberofDevelopersMetric(Graph graph, string vcsPath, string oldRevision, string newRevision)
+        {
+            using (var repo = new Repository(vcsPath))
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:186]: [Do not use `var`.](https://github.com/uni-bremen-agst/SEE/wiki/ReviewChecklist)

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+            {
+                var oldCommit = repo.Lookup<Commit>(oldRevision);
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:188]: [Do not use `var`.](https://github.com/uni-bremen-agst/SEE/wiki/ReviewChecklist)

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+                var newCommit = repo.Lookup<Commit>(newRevision);
+
+                var commits = repo.Commits.QueryBy(new CommitFilter { SortBy = CommitSortStrategies.Topological });
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:191]: [Do not use `var`.](https://github.com/uni-bremen-agst/SEE/wiki/ReviewChecklist)

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 11]
```diff
@@ -0,0 +1,282 @@
...
+
+                var uniqueContributorsPerFile = new Dictionary<string, HashSet<string>>();
+
+                foreach (var commit in commits)
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:195]: [Do not use `var`.](https://github.com/uni-bremen-agst/SEE/wiki/ReviewChecklist)

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 12]
```diff
@@ -0,0 +1,282 @@
...
+                {
+                    if (commit.Author.When >= oldCommit.Author.When && commit.Author.When <= newCommit.Author.When)
+                    {
+                        foreach (var parent in commit.Parents)
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:199]: [Do not use `var`.](https://github.com/uni-bremen-agst/SEE/wiki/ReviewChecklist)

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 13]
```diff
@@ -0,0 +1,282 @@
...
+                        {
+                            var changes = repo.Diff.Compare<Patch>(parent.Tree, commit.Tree);
+
+                            foreach (var change in changes)
+                            {
+                                var filePath = change.Path;
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:205]: [Do not use `var`.](https://github.com/uni-bremen-agst/SEE/wiki/ReviewChecklist)

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+                                var id = commit.Author.Email;
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:206]: [Do not use `var`.](https://github.com/uni-bremen-agst/SEE/wiki/ReviewChecklist)

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 15]
```diff
@@ -0,0 +1,282 @@
...
+                                {
+                                    uniqueContributorsPerFile[filePath] = new HashSet<string>();
+                                }
+
+                                uniqueContributorsPerFile[filePath].Add(id);
+                            }
+                        }
+                    }
+                }
+                foreach (var entry in uniqueContributorsPerFile)
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:218]: [Do not use `var`.](https://github.com/uni-bremen-agst/SEE/wiki/ReviewChecklist)

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 16]
```diff
@@ -0,0 +1,282 @@
...
+        /// </summary>
+        /// <param name="graph">an existing graph where to add the metrics</param>
+        /// <param name="vcsPath">the path to the VCS containing the two revisions to be compared</param>
+        /// <param name="oldRevision">the older revision that constitutes the baseline of the comparison</param>
+        /// <param name="newRevision">the newer revision against which the <see cref="oldRevision"/> is to be compared</param>
+        protected void AddCommitFrequencyMetric(Graph graph, string vcsPath, string oldRevision, string newRevision)
+        {
+            using (var repo = new Repository(vcsPath))
+            {
+                var oldCommit = repo.Lookup<Commit>(oldRevision);
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:242]: [Do not use `var`.](https://github.com/uni-bremen-agst/SEE/wiki/ReviewChecklist)

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 17]
```diff
@@ -0,0 +1,282 @@
...
+                var newCommit = repo.Lookup<Commit>(newRevision);
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:243]: [Do not use `var`.](https://github.com/uni-bremen-agst/SEE/wiki/ReviewChecklist)

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 18]
```diff
@@ -0,0 +1,282 @@
...
+
+                var commitsBetween = repo.Commits.QueryBy(new CommitFilter
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:245]: [Do not use `var`.](https://github.com/uni-bremen-agst/SEE/wiki/ReviewChecklist)

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 19]
```diff
@@ -0,0 +1,282 @@
...
+                {
+                    IncludeReachableFrom = newCommit,
+                    ExcludeReachableFrom = oldCommit
+                });
+
+                Dictionary<string, int> fileCommitCounts = new Dictionary<string, int>();
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:251]: You can use the simpler `new()` operator.

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 20]
```diff
@@ -0,0 +1,282 @@
...
+
+                foreach (var commit in commitsBetween)
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:253]: [Do not use `var`.](https://github.com/uni-bremen-agst/SEE/wiki/ReviewChecklist)

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 21]
```diff
@@ -0,0 +1,282 @@
...
+                {
+                    foreach (var parent in commit.Parents)
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:255]: [Do not use `var`.](https://github.com/uni-bremen-agst/SEE/wiki/ReviewChecklist)

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 22]
```diff
@@ -0,0 +1,282 @@
...
+                    {
+                        var changes = repo.Diff.Compare<TreeChanges>(parent.Tree, commit.Tree);
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:257]: [Do not use `var`.](https://github.com/uni-bremen-agst/SEE/wiki/ReviewChecklist)

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 23]
```diff
@@ -0,0 +1,282 @@
...
+                        foreach (var change in changes)
+                        {
+                            var filePath = change.Path;
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:260]: [Do not use `var`.](https://github.com/uni-bremen-agst/SEE/wiki/ReviewChecklist)

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 24]
```diff
@@ -0,0 +1,282 @@
...
+                            if (fileCommitCounts.ContainsKey(filePath))
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:261]: All code in bodies of control statement must be nested in curly brackets.

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 25]
```diff
@@ -0,0 +1,282 @@
...
+                                fileCommitCounts[filePath]++;
+                            else
+                                fileCommitCounts.Add(filePath, 1);
+                        }
+                    }
+                }
+
+                foreach (var entry in fileCommitCounts.OrderByDescending(x => x.Value))
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:269]: [Do not use `var`.](https://github.com/uni-bremen-agst/SEE/wiki/ReviewChecklist)

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 26]
```diff
@@ -0,0 +1,282 @@
...
+                {
+                    foreach (var node in graph.Nodes())
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:271]: [Do not use `var`.](https://github.com/uni-bremen-agst/SEE/wiki/ReviewChecklist)

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 27]
```diff
@@ -0,0 +1,282 @@
...
+                    {
+                        if (node.ID.Replace('\\', '/') == entry.Key)
+                        {
+                            node.SetInt("Commit Frequency", entry.Value);
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:275]: Metric names must start with prefix `Metric.` and must not contain blanks.

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 28]
```diff
@@ -0,0 +1,282 @@
...
+                        }
+                    }
+                }
+                foreach (var entry in uniqueContributorsPerFile)
+                {
+                    foreach (var node in graph.Nodes())
+                    {
+                        if (node.ID.Replace('\\', '/') == entry.Key)
+                        {
+                            node.SetInt("Number of Developers", entry.Value.Count);
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:224]: Metric names must start with prefix `Metric.` and must not contain blanks.

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 29]
```diff
@@ -0,0 +1,282 @@
...
+
+                var changes = repo.Diff.Compare<Patch>(oldCommit.Tree, newCommit.Tree);
+
+                foreach (var change in changes)
+                {
+                    foreach (var node in graph.Nodes())
+                    {
+                        if (node.ID.Replace('\\', '/') == change.Path)
+                        {
+                            node.SetInt("Lines Added", change.LinesAdded);
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:169]: Metric names must start with prefix `Metric.` and must not contain blanks.

File: Assets/SEE/GraphProviders/GitGraphProvider.cs [hunk 30]
```diff
@@ -0,0 +1,282 @@
...
+                            node.SetInt("Lines Deleted", change.LinesDeleted);
```
Comment (koschke) [inline Assets/SEE/GraphProviders/GitGraphProvider.cs:170]: Metric names must start with prefix `Metric.` and must not contain blanks.

**Review by koschke — CHANGES_REQUESTED** [review]
Overall, this looks good. I have only minor change requests.

- Do not use 'var' (except for rare exceptions where not using it would render the code less readable).
- Metric names must start with prefix `Metric.` and must not contain blanks. Consider to define the names in `SEE.DataModel.DG.StandardNames.cs`

**Comment by Omidious** [comment]
I applied the changes requested by the reviewers.

**Review by koschke — CHANGES_REQUESTED** [review]
Looks good.

Could you also provide test cases? Thanks!

File: Assets/SEE/GraphProviders/GitGraphProvider.cs (hunk 1 above)
Comment (koschke) [reply Assets/SEE/GraphProviders/GitGraphProvider.cs]: Was fixed.

File: Assets/SEE/GraphProviders/GitGraphProvider.cs (hunk 2 above)
Comment (koschke) [reply Assets/SEE/GraphProviders/GitGraphProvider.cs]: Was fixed.

File: Assets/SEE/GraphProviders/GitGraphProvider.cs (hunk 4 above)
Comment (koschke) [reply Assets/SEE/GraphProviders/GitGraphProvider.cs]: Was fixed.

File: Assets/SEE/GraphProviders/GitGraphProvider.cs (hunk 5 above)
Comment (koschke) [reply Assets/SEE/GraphProviders/GitGraphProvider.cs]: Was fixed.

File: Assets/SEE/GraphProviders/GitGraphProvider.cs (hunk 27 above)
Comment (koschke) [reply Assets/SEE/GraphProviders/GitGraphProvider.cs]: Was fixed.

File: Assets/SEE/GraphProviders/GitGraphProvider.cs (hunk 26 above)
Comment (koschke) [reply Assets/SEE/GraphProviders/GitGraphProvider.cs]: Was fixed.

File: Assets/SEE/GraphProviders/GitGraphProvider.cs (hunk 29 above)
Comment (koschke) [reply Assets/SEE/GraphProviders/GitGraphProvider.cs]: Was fixed.

File: Assets/SEE/GraphProviders/GitGraphProvider.cs (hunk 30 above)
Comment (koschke) [reply Assets/SEE/GraphProviders/GitGraphProvider.cs]: Was fixed.

File: Assets/SEE/GraphProviders/GitGraphProvider.cs (hunk 3 above)
Comment (koschke) [reply Assets/SEE/GraphProviders/GitGraphProvider.cs]: Was fixed.

File: Assets/SEE/GraphProviders/GitGraphProvider.cs (hunk 6 above)
Comment (koschke) [reply Assets/SEE/GraphProviders/GitGraphProvider.cs]: Was fixed.

File: Assets/SEE/GraphProviders/GitGraphProvider.cs (hunk 8 above)
Comment (koschke) [reply Assets/SEE/GraphProviders/GitGraphProvider.cs]: Was fixed.

File: Assets/SEE/GraphProviders/GitGraphProvider.cs (hunk 10 above)
Comment (koschke) [reply Assets/SEE/GraphProviders/GitGraphProvider.cs]: Was fixed.

File: Assets/SEE/GraphProviders/GitGraphProvider.cs (hunk 19 above)
Comment (koschke) [reply Assets/SEE/GraphProviders/GitGraphProvider.cs]: Was fixed.

File: Assets/SEE/GraphProviders/GitGraphProvider.cs (hunk 18 above)
Comment (koschke) [reply Assets/SEE/GraphProviders/GitGraphProvider.cs]: Was fixed.

**Comment by koschke** [comment]
Please mark all suggestions you addressed as resolved.

**Comment by Omidious** [comment]
I added the requested test cases for GitGraphProvider

**Pull request merged**